fix(ToolTip): fix the issue that Tooltip does not hide when button is clicked

TopProgrammer77 opened the pull request:
GitHub Issue (If applicable): closes #10979

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal), unless the change is documentation related. -->

## PR Type

What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR

- Bugfix
- Feature
- Code style update (formatting)
- Refactoring (no functional changes, no api changes)
- Build or CI related changes
- Documentation content changes
- Project automation
- Other... Please describe:

-->
- Bugfix
## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
https://github.com/unoplatform/uno/issues/10979

## What is the new behavior?

<!-- Please describe the new behavior after your modifications. -->


## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

Xiaoy312 commented on src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTipService.cs:
```diff
@@ -167,6 +167,13 @@ private static void OnOwnerLoaded(object sender, RoutedEventArgs e)
 			{
 				owner.PointerEntered += OnPointerEntered;
 				owner.PointerExited += OnPointerExited;
+				owner.Tapped += OnTapped;
```
It is been noted that only some controls dismiss the tooltip on press, namely the clickable ones from `ButtonBase` and some other controls such as `TextBox` or `ComboBox`. Others like `ListView` or `Slider` kept the tooltip opened through clicks.

And, it is not necessarily the Tapped/KeyDown events that turn off the tooltip. For instance, `TextBox` only hides tooltip on TextChanged. Arrows keys to (which would trigger KeyDown) doesnt cause the tooltip to be hidden.

TopProgrammer77 replied:
I checked again in  Windows App.

- As you say, only `ButtonBase` close tooltip when pointer pressed. 
- And other controls including `ListView` and `Slider` close tooltip on click.
- Keyboard events (except arrows) always close Tooltip.
- There are no other cases.

@Xiaoy312 , I want your advice before editing the code.

Xiaoy312 replied:
- `ListView`: To expand on this, hovering an LVItem to have the tooltip being shown, and then clicking the same LVI does not close the tooltip, but clicking a different one will do.
- `Slider`: Tooltip are kept opened while dragging **until released**, if you start dragging from the hover spot that is away from the thumb.
- keyboard: There are some other exceptions in addition to arrow keys, like: <kbd>alt</kbd><kbd>ins/del/home/end</kbd>
  But, only if they don't change the focus. So, keeping KeyDown would be better choice, than other specific events, as it is more precise and applicable to most controls (from UIElement).

Xiaoy312 replied:
![lv slider tooltip](https://github.com/unoplatform/uno/assets/2359550/fa36180a-86c6-4c51-bac6-e1b37a38437a)

Xiaoy312 replied:
It is fine if you can't perfectly replicate the expected behaviors for LV/LVItem or Slider. You can open issue(s) about the discrepancies that remains.

Xiaoy312 commented on src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTipService.cs:
```diff
@@ -222,5 +236,38 @@ private static void OnPointerExited(object sender, PointerRoutedEventArgs e)
+		{
+			if (sender is FrameworkElement owner && GetToolTipReference(owner) is { } toolTip)
+			{
+				if (e.GetCurrentPoint(owner).Properties.IsLeftButtonPressed)
+					toolTip.IsOpen = false;
```
```suggestion
				if (e.GetCurrentPoint(owner).Properties.IsLeftButtonPressed)
				{
					toolTip.IsOpen = false;
				}
```

TopProgrammer77 commented:
[CanChangeDisplayModeBeforeLoaded()](https://github.com/unoplatform/uno/pull/12307/checks?check_run_id=13526855162),
[When_FromUri()](https://github.com/unoplatform/uno/pull/12307/checks?check_run_id=13526855162)
are not related my commits.
When_FromUri() is failed even in master branch.

TopProgrammer77 commented:
Hi, @Xiaoy312 , @mterwoord .
I've tried to add the run time test but I think it was difficult. Because  it is difficult to emit PointerPressed Event Programmatically.
Can I remove this case of  the test?

TopProgrammer77 commented:
Hi, @jeromelaban . Is it okay not to add tests? If no, how should I add test?

jeromelaban commented:
There are [tests here](https://github.com/unoplatform/uno/blob/4ae0add1a1b55d93756f4c67bf86e6d050572cce/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ToolTip.cs#L308), which are validating some of the behaviors. You may not be able to simulate hovering to make the tooltip show up, but you should be able to open it through `IsOpen` then click the button to validate that it hides properly.

TopProgrammer77 commented:
Hi, @jeromelaban. Which I want to know is how to emit **PointerPressed** event programmatically. I didn't use **click event** and It is impossible to resolve the problem with `RaiseClick()`. I've already tested with `RaiseClick()`

TopProgrammer77 commented:
https://github.com/unoplatform/uno/pull/12307/commits/c40d1b519a6f556b057426322ef23c80bfe665ee

This was the test method which I added and removed again before in this PR.

jeromelaban commented:
> Hi, @jeromelaban. Which I want to know is how to emit **PointerPressed** event programmatically. I didn't use **click event** and It is impossible to resolve the problem with `RaiseClick()`. I've already tested with `RaiseClick()`.
> Instead of **Click** Event, I used **Tapped** Event and **KeyDown** Event( for Enter).

Thanks, I missed that part. @dr1rrb may be able to help there.

TopProgrammer77 commented:
> > Hi, @jeromelaban. Which I want to know is how to emit **PointerPressed** event programmatically. I didn't use **click event** and It is impossible to resolve the problem with `RaiseClick()`. I've already tested with `RaiseClick()`.
> > Instead of **Click** Event, I used **Tapped** Event and **KeyDown** Event( for Enter).
> 
> Thanks, I missed that part. @dr1rrb may be able to help there.

Hi, @dr1rrb . Could you help me this problem?

TopProgrammer77 commented:
Hi, @jeromelaban. I added the test again. I resolved it with TappedEvent. Thank you.

jeromelaban commented:
/azp run

TopProgrammer77 commented:
Hello, @jeromelaban, @Xiaoy312 . All checks have passed. Could you review my PR?

TopProgrammer77 commented:
Thank you, @dr1rrb !
